src/mlpack/methods/adaboost/adaboost_regressor_impl.hpp

```diff
@@ -0,0 +1,386 @@
+{
+  // Check edge case.
+  if (trees.size() == 0)
+  { 
+    predictions.clear();
+
+    throw std::invalid_argument("AdaBoostRegressor::Predict(): no AdaBoost Regressor "
+        "trained!");
+  }
+
+  predictions.set_size(data.n_cols);
+
+  #pragma omp parallel for
+  for (size_t i = 0; i < data.n_cols; ++i)
+  {
+    predictions[i] = Predict(data.col(i));
+  }
+}
+
+template<
+    typename LossFunctionType,
+    typename FitnessFunction,
+    typename DimensionSelectionType,
+    template<typename> class NumericSplitType,
+    template<typename> class CategoricalSplitType
+>
+template<typename Archive>
+void AdaBoostRegressor<
+    LossFunctionType,
+    FitnessFunction,
+    DimensionSelectionType,
+    NumericSplitType,
+    CategoricalSplitType
+>::serialize(Archive& ar, const uint32_t /* version */)
+{
+  size_t numTrees;
+  if (cereal::is_loading<Archive>())
+    trees.clear();
+  else
+    numTrees = trees.size();
+
+  ar(CEREAL_NVP(numTrees));
+
+  // Allocate space if needed.
+  if (cereal::is_loading<Archive>())
+    trees.resize(numTrees);
+
+  ar(CEREAL_NVP(trees));
+  ar(CEREAL_NVP(avgGain));
+}
+
+template<
+    typename LossFunctionType,
+    typename FitnessFunction,
+    typename DimensionSelectionType,
+    template<typename> class NumericSplitType,
+    template<typename> class CategoricalSplitType
+>
+template<bool UseDatasetInfo, typename MatType, typename ElemType>
+double AdaBoostRegressor<
+    LossFunctionType,
+    FitnessFunction,
+    DimensionSelectionType,
+    NumericSplitType,
+    CategoricalSplitType
+>::Train(const MatType& data,
+         const data::DatasetInfo& datasetInfo,
+         const arma::Row<ElemType>& responses,
+         const size_t numTrees,
+         const size_t minimumLeafSize,
+         const double minimumGainSplit,
+         const size_t maximumDepth,
+         DimensionSelectionType& dimensionSelector)
+{
+  // Clear trees.
+  trees.clear();
+
+  // Clear confidence of trees.
+  confidence.clear();
+
+  // Load initial weights to 1 / n_responses.
+  arma::rowvec weights(responses.n_cols);
+  weights.fill(1/(double)responses.n_cols);
+
+
+  // To be used for prediction.
+  arma::Row<ElemType> predictedValues(responses.n_cols);
+
+  double totalGain = 0.0;
+
+  for (size_t i=0; i < numTrees; ++i)
+  {
+    // avoid extremely small sample weight
+    weights.clean(arma::datum::eps);
+
+    DecisionTreeType tree;
+
+    if (UseDatasetInfo)
+    {
+      totalGain += tree.Train(data, datasetInfo, responses, weights,
+            minimumLeafSize, minimumGainSplit, maximumDepth, dimensionSelector);
+    }
+    else
+    {
+      totalGain += tree.Train(data, responses, weights, minimumLeafSize,
+            minimumGainSplit, maximumDepth, dimensionSelector);
+    }
+
+    tree.Predict(data, predictedValues);
+
+    // Calculate loss based on given lossFunction.
+    arma::Row<ElemType> diff = predictedValues - responses;
+    arma::Row<ElemType> errorVec = LossFunctionType::Calculate(diff);
```

Comment:
I'd suggest an API where you just take in both the predicted values and the responses, instead of needing to calculate the intermediate `diff` vector.